services/scan/src/precinct_scanner_state_machine.ts

```diff
@@ -0,0 +1,854 @@
+function assign(arg: Assigner<Context, any> | PropertyAssigner<Context, any>) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return xassign<Context, any>(arg);
+}
+
+type ConfigurationEvent =
+  | { type: 'CONFIGURE'; store: Store; interpreter: SimpleInterpreter }
+  | { type: 'UNCONFIGURE' };
+
+type ScannerStatusEvent =
+  | { type: 'SCANNER_NO_PAPER' }
+  | { type: 'SCANNER_READY_TO_SCAN' }
+  | { type: 'SCANNER_READY_TO_EJECT' }
+  | { type: 'SCANNER_BOTH_SIDES_HAVE_PAPER' }
+  | { type: 'SCANNER_JAM' }
+  | { type: 'SCANNER_DISCONNECTED' };
+
+interface Interpretation {
+  sheetId: Id;
+  sheet: SheetOf<PageInterpretationWithFiles>;
+}
+
+export type InterpretationResultEvent = Scan.InterpretationResult & {
+  interpretation: Interpretation;
+};
+
+type CommandEvent =
+  | { type: 'SCAN' }
+  | { type: 'ACCEPT' }
+  | { type: 'RETURN' }
+  | { type: 'START_OVER' }
+  | { type: 'CALIBRATE' };
+
+export type Event =
+  | ScannerStatusEvent
+  | InterpretationResultEvent
+  | CommandEvent
+  | ConfigurationEvent
+  | { type: 'SET_INTERPRETATION_MODE'; mode: InterpretationMode };
+
+async function connectToPlustek(): Promise<ScannerClient> {
+  debug('Connecting to plustek');
+  const plustekClient = await createClient(DEFAULT_CONFIG);
+  debug('Plustek client connected: %s', plustekClient.isOk());
+  return plustekClient.unsafeUnwrap();
+}
+
+function paperStatusToEvent(paperStatus: PaperStatus): ScannerStatusEvent {
+  switch (paperStatus) {
+    // When there's no paper in the scanner
+    case PaperStatus.NoPaperStatus:
+    case PaperStatus.VtmDevReadyNoPaper:
+      return { type: 'SCANNER_NO_PAPER' };
+    // When there's a paper held in the front
+    case PaperStatus.VtmReadyToScan:
+      return { type: 'SCANNER_READY_TO_SCAN' };
+    // When there's a paper held in the back
+    case PaperStatus.VtmReadyToEject:
+      return { type: 'SCANNER_READY_TO_EJECT' };
+    // When there's a paper held in the back and inserted in the front
+    case PaperStatus.VtmBothSideHavePaper:
+      return { type: 'SCANNER_BOTH_SIDES_HAVE_PAPER' };
+    // When there's a paper jammed in the scanner
+    case PaperStatus.Jam:
+    case PaperStatus.VtmFrontAndBackSensorHavePaperReady:
+      return { type: 'SCANNER_JAM' };
+    default:
+      throw new PrecinctScannerError(
+        'unexpected_paper_status',
+        `Unexpected paper status: ${paperStatus}`
+      );
+  }
+}
+
+function paperStatusErrorToEvent(
+  error: ScannerError | Error
+): ScannerStatusEvent {
+  if (error instanceof ClientDisconnectedError) {
+    return { type: 'SCANNER_DISCONNECTED' };
+  }
+  switch (error) {
+    // This error code happens when the paper gets blocked and tries to scan for
+    // a long time (or if you just hold onto the paper while scanning).
+    case ScannerError.SaneStatusInval:
+      return { type: 'SCANNER_JAM' };
+    case ScannerError.SaneStatusIoError:
+      return { type: 'SCANNER_DISCONNECTED' };
```

Comment:
are we sure we want to treat these errors as disconnected? In my playing with the scanner, this error appeared to be recoverable,

Reply:
Oh interesting I'm not sure. That was just the error that happened when I turned off the Plustek. I'm not sure how to distinguish a recoverable `SaneStatusIoError` from that case. Do you remember what triggered it when it was recoverable?

Reply:
I've seen it happen usually right as you try to scan. In your initial BallotQuest code, you went to checkpaperstatus on that error, and checkpaperstatus usually was able to get a status and recover from there.

Reply:
Hmm interesting. Is there any way I can easily repro this? I'm hesitant to change this logic without being able to test that my changes actually work.

One idea I have to fix it is to retry getting paper status a few times on error before sending this disconnected event.

Reply:
it can be repro'ed sometimes by holding the ballot back a little when it's trying to grab it / scan it.

To be clear, I'm fine with not fixing this now! We can optimize this later.